nucypher/characters/control/serializers.py

```diff
@@ -0,0 +1,219 @@
+import json
+from abc import ABC
+from base64 import b64decode, b64encode
+from json import JSONDecodeError
+from typing import Callable
+
+import maya
+
+import nucypher
+from nucypher.characters.control.specifications import CharacterSpecification
+
+
+class CharacterControlSerializer(ABC):
+
+    class SerializerError(ValueError):
+        """Invalid data for I/O serialization or deserialization"""
+
+
+class CharacterControlJSONSerializer(CharacterControlSerializer):
+
+    def __call__(self, data, specification: tuple, *args, **kwargs):
+        if isinstance(data, dict):
+            return self.__serialize(response_data=data, output_specification=specification)
+        elif isinstance(data, bytes):
+            return self.__deserialize(request_payload=data, input_specification=specification)
+        else:
+            error_message = f"{self.__class__.__name__} only accepts dict or bytes as input. Got {data.__class__.__name__} "
+            raise ValueError(error_message)
+
+    @staticmethod
+    def build_response_metadata(response: dict, duration: maya.timedelta) -> dict:
+        response_data = {'result': response,
+                         'version': str(nucypher.__version__),
+                         'duration': str(duration)}
+        return response_data
+
+    @staticmethod
+    def validate_input(request_data: dict, input_specification: tuple) -> bool:
+        # Handle client input
+
+        # Invalid Fields
+        input_fields = set(request_data.keys())
+        extra_fields = input_fields - set(input_specification)
+
+        if extra_fields:
+            raise CharacterSpecification.InvalidInputField(f"Invalid request fields '{', '.join(extra_fields)}'."
+                                                           f"Valid fields are: {', '.join(input_specification)}.")
+
+        # Missing Fields
+        missing_fields = list()
```

Comment:
I think, you could simply do `missing_fields = set(input_specification) - input_fields` (extra fields in input_fields will be ignored)

Reply:
This is in part used for error handling, but I do agree that is can be much more concise with sets.  I'll make an update.

Reply:
I'm thinking of opening this as an Issue instead, lots of changes happening in this PR.

Reply:
https://github.com/nucypher/nucypher/issues/813